xilinx/src/xlnx_dec_utils.rs

```diff
@@ -0,0 +1,228 @@
+            Ok(lib) => {
+                match lib
+                    .get::<Symbol<extern "C" fn(props: *mut XmaDecoderProperties, function: *const c_char, json_params: *mut c_char)>>(b"convertXmaPropsToJson")
+                {
+                    Ok(convert_xma_props_to_json) => {
+                        let function = CString::new("DECODER").unwrap();
+                        convert_xma_props_to_json(xma_dec_props, function.as_ptr(), plugin_param.input.as_mut_ptr());
+                    }
+                    Err(e) => bail!("{}", e),
+                }
+            }
+            Err(e) => bail!("{}", e),
+        };
+    }
+
+    let plugin_name = CString::new("xrmU30DecPlugin").unwrap().into_raw();
+
+    unsafe {
+        let ret = xrmExecPluginFunc(xrm_ctx, plugin_name, func_id, &mut plugin_param);
+        if ret != XRM_SUCCESS as i32 {
+            bail!("XRM decoder plugin failed to calculate decoder load. error: {}", ret);
+        }
+    }
+    // parse the load from the output buffer of plugin param.
+    let output_bytes = &plugin_param.output.map(|i| i as u8);
+    let dec_plugin_output = from_utf8(output_bytes).unwrap_or("-1").split(' ').collect::<Vec<&str>>()[0];
+    let load = dec_plugin_output.parse::<i32>().unwrap_or(-1);
+
+    if load == -1 {
+        bail!("Unable to parse load calculation from XRM decoder plugin");
+    }
+
+    Ok(load)
+}
+
+pub fn xlnx_fill_dec_pool_props(cu_pool_prop: &mut xrmCuPoolProperty, dec_load: i32) -> Result<(), SimpleError> {
+    cu_pool_prop.cuListProp.sameDevice = true;
+    cu_pool_prop.cuListNum = 1;
+
+    strcpy_to_arr_i8(&mut cu_pool_prop.cuListProp.cuProps[0].kernelName, "decoder")?;
+    strcpy_to_arr_i8(&mut cu_pool_prop.cuListProp.cuProps[0].kernelAlias, "DECODER_MPSOC")?;
+    cu_pool_prop.cuListProp.cuProps[0].devExcl = false;
+    cu_pool_prop.cuListProp.cuProps[0].requestLoad = xrm_precision_1000000_bitmask(dec_load);
+
+    strcpy_to_arr_i8(&mut cu_pool_prop.cuListProp.cuProps[1].kernelName, "kernel_vcu_decoder")?;
+
+    cu_pool_prop.cuListProp.cuProps[1].devExcl = false;
+    cu_pool_prop.cuListProp.cuProps[1].requestLoad = xrm_precision_1000000_bitmask(XRM_MAX_CU_LOAD_GRANULARITY_1000000 as i32);
+    // we defined 2 cu requests to the properties.
+    cu_pool_prop.cuListProp.cuNum = 2;
+
+    Ok(())
+}
+
+pub fn xlnx_reserve_dec_resource(xlnx_dec_ctx: &mut XlnxDecoderXrmCtx) -> Result<(), SimpleError> {
+    // a device has already been chosen, there is no need to assign a reserve id.
+    if xlnx_dec_ctx.device_id >= 0 {
+        return Ok(());
+    }
+
+    let mut cu_pool_prop: xrmCuPoolProperty = Default::default();
+    xlnx_fill_dec_pool_props(&mut cu_pool_prop, xlnx_dec_ctx.dec_load)?;
+
+    unsafe {
+        let num_cu_pool = xrmCheckCuPoolAvailableNum(xlnx_dec_ctx.xrm_ctx, &mut cu_pool_prop);
+        if num_cu_pool <= 0 {
+            bail!("No decoder resources available for allocation")
```

Comment:
Noticing inconsistent caps on these errors. Typically we use lowercase errors:

```suggestion
            bail!("no decoder resources available for allocation")
```